zapcore/field.go

```diff
@@ -151,7 +151,13 @@ func (f Field) AddTo(enc ObjectEncoder) {
 	case StringerType:
 		enc.AddString(f.Key, f.Interface.(fmt.Stringer).String())
 	case ErrorType:
-		enc.AddString(f.Key, f.Interface.(error).Error())
+		val := f.Interface.(error)
+		if fancy, ok := val.(fmt.Formatter); ok {
+			// This is a rich error type, like those produced by
+			// github.com/pkg/errors.
```

Comment:
Suggest verifying that the verbose error is actually different from the error message before adding.

It won't cost much extra (since the string length check will probably fail first)